### DISCUSSION TITLE
DB Housekeeping

### DISCUSSION THREAD

**Discussion by mityakoval** [ory/hydra#2556] [Q&A] [answered]
Hi! 

Continuing on the DB housekeeping topic from the community website.
We are currently trying to upgrade from `v1.8.5` to `v1.10.2` which includes creating indeces on:

* `hydra_oauth2_access`
* `hydra_oauth2_refresh`
* `hydra_oauth2_code`
* `hydra_oauth2_oidc`
* `hydra_oauth2_pkce`

While `/oauth2/flush` cleans up the first table, we have an even bigger problem with `hydra_oauth2_code` (index creation took about 9 hours) and we can't allow this much downtime.

What would be the strategy to clean this table up?

**Comment by vinckr** [comment, marked as answer]
Have you had a look at `hydra janitor` ?

 https://www.ory.sh/hydra/docs/cli/hydra-janitor/

**Reply by mityakoval** [reply]
Yes, but it's only available from v1.10, and we are currently on 1.8.5

**Reply by vinckr** [reply]
Oh sorry, I see you are upgrading at the moment. 
I will do some research if I can find someone who solved a similar issue!

**Reply by mityakoval** [reply]
Thank you! It would be much appreciated! 

I found these two on the community website:
* https://community.ory.sh/t/online-db-migration/1890
* https://community.ory.sh/t/db-housekeeping/2224

But there was no actual solution for the problem

**Reply by aeneasr** [reply]
You could try running the janitor from 1.10 against a back up of your database and see if that helps. The data model hasn't really changed a lot I believe. Use with caution though (and not against production) :)

**Reply by Benehiko** [reply]
Hey @mityakoval,

This all depends on your constraints/limitations in your current setup. For instance, if it is possible to invalidate tokens or if the tokens you are creating have a very short lifetime this might make things easier. 

I would think that since you have a really large database, that it would be difficult to avoid downtime with a migration to the latest version of Hydra without make database alterations (e.g. deleting data). I can think of two ways to go about this, 1) Run two instances and slowly port current active tokens (through invalidation / expiration) to the new instance. 2) delete old data and then do the migration

1)
It might be better to deploy the latest version of Hydra on a new clean database and redirect any new login requests to the new hydra instance. This would require you to manually port over the clients you have created from the old hydra instance to the new one. This would, however, require you to setup a "fallback" mechanism which means the old hydra instance would be requested first when checking the validity of a token and on failure it would check the new hydra instance. In my opinion, this option might be a bit over critical on trying to "save" old data which might not mean much to you. 

2)
Another option would be to do a manual delete old data not being used anymore, this isn't as difficult as you might think - but could take some time to research the correct constraints to add to the delete query. Once you have cleaned up the data you can just run the migration. Please check out [this PR](https://github.com/ory/hydra/issues/2514) for more information on the cleanup procedure you can pursue.

**Reply by mityakoval** [reply]
Thanks a lot for your replies guys! 

---

Regarding the manual clean up of `hydra_oauth2_code`, obvious constraints would be `active=false` and `requested_at<*some_threshold*`. Would that be sufficient in your opinion? 
We will certainly be creating some batch jobs to clean all of the tables, but judging by the size of this table deleting all unused data would be a quite time consuming task in itself.

---

Since `hydra token flush` clears access token, is it able to clean up the auth codes table cascadingly? 

---

Also one option we are considering is to run those migrations manually. So these are the migrations:
```
DROP INDEX hydra_oauth2_access_request_id_idx;
DROP INDEX hydra_oauth2_refresh_request_id_idx;

CREATE INDEX hydra_oauth2_access_request_id_idx ON hydra_oauth2_access (request_id);
CREATE INDEX hydra_oauth2_refresh_request_id_idx ON hydra_oauth2_refresh (request_id);

CREATE INDEX hydra_oauth2_code_request_id_idx ON hydra_oauth2_code (request_id);
CREATE INDEX hydra_oauth2_oidc_request_id_idx ON hydra_oauth2_oidc (request_id);
CREATE INDEX hydra_oauth2_pkce_request_id_idx ON hydra_oauth2_pkce (request_id);
```

Our idea is:
1. Re-create indices in `hydra_oauth2_access` and `hydra_oauth2_refresh` as those seem to complete in under an hour
2. Run the creation of the remaining indices `CONCURRENTLY` (we're using PostgreSQL 10) on a live DB
3. Then bump up the schema version manually
4. Proceed with the remaining migration through `hydra migrate`, as it takes no time to complete

**Reply by mityakoval** [reply]
> You could try running the janitor from 1.10 against a back up of your database and see if that helps. The data model hasn't really changed a lot I believe. Use with caution though (and not against production) :)

I don't see how `janitor` is removing auth codes. Token and login/consent requests, but not the codes explicitly. Does it do it cascadingly?

**Reply by aeneasr** [reply]
Removing auth codes will probably invalidate access tokens bound to that code so there will be some effort into finding out which auth codes can actually be removed, using some joins for example.

**Reply by Benehiko** [reply]
> I don't see how janitor is removing auth codes. Token and login/consent requests, but not the codes explicitly. Does it do it cascadingly?

Yes it doesn't remove auth codes currently, and doesn't cascade. This will need to be done manually.
